### ISSUE TITLE
cargo run doesn't work

### ISSUE THREAD

**Issue by GraydenH** [phil-opp/blog_os#1065]
hi, according to the readme cargo run is supposed to run the bin file generated by cargo bootimage using qemu.

I get an error if I run cargo bootimage
```
cargo bootimage
WARNING: `CARGO_MANIFEST_DIR` env variable not set
Building kernel
warning: unused manifest key: build
warning: unused manifest key: unstable
   Compiling os v0.1.0 (C:\Users\grayd\IdeaProjects\os)
error: language item required, but not found: `eh_personality`

error: could not compile `os` due to previous error
Error: Kernel build failed.
Stderr:
```

when i run `cargo run -Zbuild-std  --target x86_64-os.json` I get the expected output.

This is the error i get after running cargo run, now that i have run cargo bootimage successfully and created the .bin file.

```
cargo run
warning: unused manifest key: build
   Compiling os v0.1.0 (C:\Users\grayd\IdeaProjects\os)
error: language item required, but not found: `eh_personality`

error: could not compile `os` due to previous error
```

this is the following cargo.toml

```
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[build]
target = "x86_64-os.json"

[unstable]
build-std-features = ["compiler-builtins-mem"]
build-std = ["core", "compiler_builtins"]

[dependencies]
bootloader = "0.9.8"
```

I think i followed all the instructions. I did also have to run `rustup component add llvm-tools-preview`, i think to run cargo bootimage.

My code is at https://github.com/GraydenH/os if you want to try reproduce it,

**Comment by GraydenH** [comment]
This is after completing post 2 and trying to run it

**Comment by bjorn3** [comment]
The build and unstable sections need to go into `.cargo/config.toml` and not `Cargo.toml`.

**Comment by GraydenH** [comment]
@bjorn3 Hi, now i get a new error. I deleted my target folder and re ran build, bootimage and run. Build and bootimage were fine but run doesn't work.

```
C:\Users\grayd\IdeaProjects\os>cargo build
   Compiling compiler_builtins v0.1.52
   Compiling core v0.0.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core)
   Compiling bootloader v0.9.19
   Compiling rustc-std-workspace-core v1.99.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\rustc-std-workspace-core)
   Compiling os v0.1.0 (C:\Users\grayd\IdeaProjects\os)
    Finished dev [unoptimized + debuginfo] target(s) in 28.72s

C:\Users\grayd\IdeaProjects\os>cargo bootimage
WARNING: `CARGO_MANIFEST_DIR` env variable not set
Building kernel
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
Building bootloader
   Compiling compiler_builtins v0.1.52
   Compiling core v0.0.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core)
   Compiling serde v1.0.116
   Compiling bitflags v1.2.1
   Compiling llvm-tools v0.1.1
   Compiling x86_64 v0.13.2
   Compiling toml v0.5.6
   Compiling bootloader v0.9.19 (C:\Users\grayd\.cargo\registry\src\github.com-1ecc6299db9ec823\bootloader-0.9.19)
   Compiling rustc-std-workspace-core v1.99.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\rustc-std-workspace-core)
   Compiling bit_field v0.9.0
   Compiling zero v0.1.2
   Compiling usize_conversions v0.2.0
   Compiling rlibc v1.0.0
   Compiling fixedvec v0.2.4
   Compiling xmas-elf v0.6.2
warning: use of deprecated macro `llvm_asm`: will be removed from the compiler, use asm! instead
  --> src\main.rs:45:5
   |
45 |     llvm_asm!("call $1; ${:private}.spin.${:uid}: jmp ${:private}.spin.${:uid}" ::
   |     ^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: use of deprecated macro `llvm_asm`: will be removed from the compiler, use asm! instead
  --> src\main.rs:91:5
   |
91 |     llvm_asm!("mov bx, 0x0
   |     ^^^^^^^^

warning: `bootloader` (bin "bootloader") generated 2 warnings
    Finished release [optimized + debuginfo] target(s) in 37.85s
Created bootimage for `os` at `C:\Users\grayd\IdeaProjects\os\target\x86_64-os\debug\bootimage-os.bin`

C:\Users\grayd\IdeaProjects\os>cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target\x86_64-os\debug\os`
error: could not execute process `target\x86_64-os\debug\os` (never executed)

Caused by:
  %1 is not a valid Win32 application. (os error 193)
```

**Comment by bjorn3** [comment]
You are missing

```toml
[target.'cfg(target_os = "none")']
runner = "bootimage runner"
```

in `.cargo/config.toml`. These lines cause `cargo run` to run the produced kernel inside a virtual machine instead of directly as executable.

**Comment by GraydenH** [comment]
Thanks,

I get a new error after cleaning and rerunning

```

C:\Users\grayd\IdeaProjects\os>cargo build
   Compiling compiler_builtins v0.1.52
   Compiling core v0.0.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core)
   Compiling bootloader v0.9.19
   Compiling rustc-std-workspace-core v1.99.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\rustc-std-workspace-core)
   Compiling os v0.1.0 (C:\Users\grayd\IdeaProjects\os)
    Finished dev [unoptimized + debuginfo] target(s) in 21.85s

C:\Users\grayd\IdeaProjects\os>cargo bootimage
WARNING: `CARGO_MANIFEST_DIR` env variable not set
Building kernel
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
Building bootloader
   Compiling compiler_builtins v0.1.52
   Compiling core v0.0.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core)
   Compiling serde v1.0.116
   Compiling bitflags v1.2.1
   Compiling llvm-tools v0.1.1
   Compiling x86_64 v0.13.2
   Compiling toml v0.5.6
   Compiling bootloader v0.9.19 (C:\Users\grayd\.cargo\registry\src\github.com-1ecc6299db9ec823\bootloader-0.9.19)
   Compiling rustc-std-workspace-core v1.99.0 (C:\Users\grayd\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\rustc-std-workspace-core)
   Compiling zero v0.1.2
   Compiling bit_field v0.9.0
   Compiling fixedvec v0.2.4
   Compiling usize_conversions v0.2.0
   Compiling rlibc v1.0.0
   Compiling xmas-elf v0.6.2
warning: use of deprecated macro `llvm_asm`: will be removed from the compiler, use asm! instead
  --> src\main.rs:45:5
   |
45 |     llvm_asm!("call $1; ${:private}.spin.${:uid}: jmp ${:private}.spin.${:uid}" ::
   |     ^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: use of deprecated macro `llvm_asm`: will be removed from the compiler, use asm! instead
  --> src\main.rs:91:5
   |
91 |     llvm_asm!("mov bx, 0x0
   |     ^^^^^^^^

warning: `bootloader` (bin "bootloader") generated 2 warnings
    Finished release [optimized + debuginfo] target(s) in 38.83s
Created bootimage for `os` at `C:\Users\grayd\IdeaProjects\os\target\x86_64-os\debug\bootimage-os.bin`

C:\Users\grayd\IdeaProjects\os>cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `bootimage runner target\x86_64-os\debug\os`
Building bootloader
   Compiling bootloader v0.9.19 (C:\Users\grayd\.cargo\registry\src\github.com-1ecc6299db9ec823\bootloader-0.9.19)
warning: use of deprecated macro `llvm_asm`: will be removed from the compiler, use asm! instead
  --> src\main.rs:45:5
   |
45 |     llvm_asm!("call $1; ${:private}.spin.${:uid}: jmp ${:private}.spin.${:uid}" ::
   |     ^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: use of deprecated macro `llvm_asm`: will be removed from the compiler, use asm! instead
  --> src\main.rs:91:5
   |
91 |     llvm_asm!("mov bx, 0x0
   |     ^^^^^^^^

warning: `bootloader` (bin "bootloader") generated 2 warnings
    Finished release [optimized + debuginfo] target(s) in 2.69s
Running: `qemu-system-x86_64 -drive format=raw,file=target\x86_64-os\debug\bootimage-os.bin`
Error: Failed to execute QEMU run command `"qemu-system-x86_64" "-drive" "format=raw,file=target\\x86_64-os\\debug\\bootimage-os.bin"`: An I/O error occured: program not found
error: process didn't exit successfully: `bootimage runner target\x86_64-os\debug\os` (exit code: 1)


```

**Comment by GraydenH** [comment]
i see that `target\\x86_64-os\\debug\\bootimage-os.bin` does exist so im not sure why it complains

**Comment by bjorn3** [comment]
You need to install qemu and have it in your `PATH`.

**Comment by GraydenH** [comment]
It runs now but the screen is black. When i run the post-2 branch with cargo run i get  the same thing. Is something about my environment wrong or is it something else?

**Comment by bjorn3** [comment]
Which qemu version did you install? I believe there is one specific version that has a bug causing the screen to go blank for blog os.

**Comment by GraydenH** [comment]
I have `6.0.93` I think it was the latest because i just got it off the website this week. What versions have the bug? I'd like to know which version i can use that works. Maybe it should be added to the guide/readme

**Comment by GraydenH** [comment]
it worked with the 2021-02-01 release i found here https://qemu.weilnetz.de/w64/2020/

**Comment by bjorn3** [comment]
The bug is in qemu 6.0 which you used. It is fixed in 6.1: https://github.com/phil-opp/blog_os/issues/1052#issuecomment-924303421

**Comment by phil-opp** [comment]
I think this is solved and can be closed. Thanks @bjorn3!
